numba/cuda/compiler.py

```diff
@@ -1123,6 +1026,40 @@ def compile(self, sig):
             self.sigs.append(sig)
         return kernel
 
+    def compile_device(self, args):
+        """Compile the function for the given argument types.
+
+        Each signature is compiled once by caching the compiled function inside
+        this object.
+
+        Returns the `CompileResult`.
+        """
+        if args not in self.overloads:
+            opt = self.targetoptions.get('opt', True)
+            debug = self.targetoptions.get('debug', False)
+            nvvm_options = {
+                'opt': 3 if opt else 0,
+                'debug': debug,
+            }
+
+            cres = compile_cuda(self.py_func, None, args, debug=debug,
```

Comment:
Why does the `compile_cuda` call not pass `device=True`? (Presumably because we never use the generated PTX, so it does not matter that NVVM doesn't produce a device function here?)

Reply:
(in which case, it may not matter whether something got compiled as a kernel first then used as a device function (as per  comment above) - perhaps then we wouldn't need separate `compile` / `compile_device` functions, and they could be combined to do everything necessary for kernels and device functions without causing issues).